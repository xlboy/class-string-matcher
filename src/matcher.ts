import {
  htmlLexer,
  jsxLexer,
  solidLexer,
  vueDynamicLexer,
  vuePureLexer,
} from './lexer';
import moo from 'moo';

const SUPPORTED_LANGUAGE_IDS = ['jsx', 'html', 'vue', 'solid'] as const;
export type LanguageId = (typeof SUPPORTED_LANGUAGE_IDS)[number];

export interface ClassNode {
  text: string;
  pos: { s: number; e: number };
}

interface VueOptions {
  /**
   * Indicates whether the text is within a dynamic context where expressions are allowed.
   * This includes scenarios like:
   * - Vue `:class="..."`
   * - Or similar constructs in other frameworks where **dynamic values** can be used for class names.
   */
  inDynamicContext?: boolean;
}

export function classStringMatcher(
  text: string,
  languageId: 'vue',
  options: VueOptions,
): ClassNode[];
export function classStringMatcher(
  text: string,
  languageId: Exclude<LanguageId, 'vue'>,
): ClassNode[];
export function classStringMatcher(
  text: string,
  languageId: LanguageId,
  options?: VueOptions,
): ClassNode[] {
  if (!SUPPORTED_LANGUAGE_IDS.includes(languageId))
    throw new Error(`Unknown languageId: ${languageId}`);

  let lexer: moo.Lexer;
  switch (languageId) {
    case 'jsx':
      lexer = jsxLexer;
      break;
    case 'vue':
      lexer = options?.inDynamicContext ? vueDynamicLexer : vuePureLexer;
      break;
    case 'solid':
      lexer = solidLexer;
      break;
    case 'html':
    default:
      lexer = htmlLexer;
      break;
  }
  lexer.reset(text);

  const parsed = parser(lexer);
  const classNodes: ClassNode[] = parsed.map(v => {
    if (v.isEmpty) {
      return {
        text: '',
        pos: { s: v.tokens[1].offset, e: v.tokens[1].offset },
      };
    }
    return {
      text: v.tokens.map(t => t.text).join(''),
      pos: { s: v.tokens[0].offset, e: v.tokens.at(-1)!.offset },
    };
  });

  return classNodes;
}

function parser(lexer: moo.Lexer) {
  const classlist: Array<{ tokens: moo.Token[]; isEmpty: boolean }> = [];
  let currentParts: moo.Token[] = [];

  // 加 try 是为了让后面无效 token 不会报错（但前面的有效 token 会被保留）
  try {
    let t: moo.Token;
    let lastT: moo.Token | undefined;
    while ((t = lexer.next()!)) {
      switch (t.type) {
        case 'ignore':
        case 'start':
          break;

        case 'children':
        case 'end': {
          if (currentParts.length > 0) {
            classlist.push({ tokens: currentParts, isEmpty: false });
          } else if (lastT?.text === t.text) {
            classlist.push({ tokens: [lastT, t], isEmpty: true });
          }
          currentParts = [];
          break;
        }

        case 'content': {
          currentParts.push(t);
          break;
        }

        default:
          throw new Error(`Unexpected token: ${t.type}`);
      }
      lastT = t;
    }
  } catch (_) {}

  return classlist;
}
