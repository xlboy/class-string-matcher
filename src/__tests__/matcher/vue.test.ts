import { classStringMatcher } from '../../matcher';
import { describe, expect, it } from 'vitest';

// describe('dynamic', () => {
//   console.log('please see jsx.test.ts');
// });

describe('pure', () => {
  const codes = [
    `"">`,
    `"text-red text-lg"`,
    `'text-blue  ...'`,
    `'a b c d ', ../...`,
  ];

  it.each(codes)('%s', code => {
    expect(
      classStringMatcher(code, 'vue', { inDynamicContext: false }),
    ).toMatchSnapshot();
  });
});
