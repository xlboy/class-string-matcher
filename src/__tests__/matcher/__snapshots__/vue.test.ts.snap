// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dynamic > ""> 1`] = `
[
  {
    "pos": {
      "e": 1,
      "s": 1,
    },
    "text": "",
  },
]
`;

exports[`dynamic > "'a b c d '", ../... 1`] = `
[
  {
    "pos": {
      "e": 9,
      "s": 2,
    },
    "text": "a b c d ",
  },
]
`;

exports[`dynamic > "'text-red' + 'text-lg'" 1`] = `
[
  {
    "pos": {
      "e": 9,
      "s": 2,
    },
    "text": "text-red",
  },
  {
    "pos": {
      "e": 21,
      "s": 15,
    },
    "text": "text-lg",
  },
]
`;

exports[`dynamic > "['1', '2', '3', \`test1 \${'test2'}\`]" 1`] = `
[
  {
    "pos": {
      "e": 3,
      "s": 3,
    },
    "text": "1",
  },
  {
    "pos": {
      "e": 8,
      "s": 8,
    },
    "text": "2",
  },
  {
    "pos": {
      "e": 13,
      "s": 13,
    },
    "text": "3",
  },
  {
    "pos": {
      "e": 23,
      "s": 18,
    },
    "text": "test1 ",
  },
  {
    "pos": {
      "e": 31,
      "s": 27,
    },
    "text": "test2",
  },
]
`;

exports[`dynamic > '"text-blue" ...' 1`] = `
[
  {
    "pos": {
      "e": 10,
      "s": 2,
    },
    "text": "text-blue",
  },
]
`;

exports[`dynamic > ('text-red', 'text-lg') 1`] = `
[
  {
    "pos": {
      "e": 9,
      "s": 2,
    },
    "text": "text-red",
  },
  {
    "pos": {
      "e": 20,
      "s": 14,
    },
    "text": "text-lg",
  },
]
`;

exports[`dynamic > \`text-ggg\` 1`] = `
[
  {
    "pos": {
      "e": 8,
      "s": 1,
    },
    "text": "text-ggg",
  },
]
`;

exports[`pure > ""> 1`] = `
[
  {
    "pos": {
      "e": 1,
      "s": 1,
    },
    "text": "",
  },
]
`;

exports[`pure > "text-red text-lg" 1`] = `
[
  {
    "pos": {
      "e": 16,
      "s": 1,
    },
    "text": "text-red text-lg",
  },
]
`;

exports[`pure > 'a b c d ', ../... 1`] = `
[
  {
    "pos": {
      "e": 8,
      "s": 1,
    },
    "text": "a b c d ",
  },
]
`;

exports[`pure > 'text-blue  ...' 1`] = `
[
  {
    "pos": {
      "e": 14,
      "s": 1,
    },
    "text": "text-blue  ...",
  },
]
`;
